### PR TITLE
Add parametrized trim functions

### PR DESCRIPTION
**Pull request by kokosing**
Add parametrized trim functions

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 1]
```diff
@@ -448,6 +449,88 @@ public static Slice trim(@SqlType("varchar(x)") Slice slice)
         return SliceUtf8.trim(slice);
     }
 
+    @Description("remove the longest string containing only given characters from the beginning of a string")
+    @ScalarFunction("ltrim")
+    @LiteralParameters("x")
+    @SqlType("varchar(x)")
+    public static Slice leftTrim(@SqlType("varchar(x)") Slice slice, @SqlType(StandardTypes.VARCHAR) Slice charactersToTrim)
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: array(varchar(1))? In future - array(char(1))?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 2]
```diff
@@ -448,6 +449,88 @@ public static Slice trim(@SqlType("varchar(x)") Slice slice)
...
+    {
+        int[] codePointsToTrim = toCodePoints(charactersToTrim);
+        int start = leftTrimPosition(slice, codePointsToTrim);
+        return slice.slice(start, slice.length() - start);
+    }
+
+    private static int leftTrimPosition(@SqlType("varchar(x)") Slice slice, int[] codePointsToTrim)
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: I think we do not want to annotate internal implementation with @SqlType annotation.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 3]
```diff
@@ -448,6 +449,88 @@ public static Slice trim(@SqlType("varchar(x)") Slice slice)
...
+    {
+        int start = 0;
+        SLICE_LOOP:
+        while (start < slice.length()) {
+            int codePoint = getCodePointAt(slice, start);
+            for (int codePointToTrim : codePointsToTrim) {
+                if (codePoint == codePointToTrim) {
+                    start += lengthOfCodePoint(start);
+                    continue SLICE_LOOP;
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: Isn't bool flag is more used around the code than 'goto'?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 4]
```diff
@@ -448,6 +449,88 @@ public static Slice trim(@SqlType("varchar(x)") Slice slice)
...
+        int end = rightTrimPosition(slice, codePointsToTrim);
+        return slice.slice(0, end);
+    }
+
+    private static int rightTrimPosition(@SqlType("varchar(x)") Slice slice, int[] codePointsToTrim)
+    {
+        int end = slice.length();
+        SLICE_LOOP:
+        while (end > 0) {
+            int codePoint = SliceUtf8.getCodePointBefore(slice, end);
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: It is probably possible to use SliceUtf8 class in leftTrimPosition as well, maybe we should do that, so the function can look more similar to each other?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 5]
```diff
@@ -448,6 +449,88 @@ public static Slice trim(@SqlType("varchar(x)") Slice slice)
...
+    @LiteralParameters("x")
+    @SqlType("varchar(x)")
+    public static Slice rightTrim(@SqlType("varchar(x)") Slice slice, @SqlType(StandardTypes.VARCHAR) Slice charactersToTrim)
+    {
+        int[] codePointsToTrim = toCodePoints(charactersToTrim);
+        int end = rightTrimPosition(slice, codePointsToTrim);
+        return slice.slice(0, end);
+    }
+
+    private static int rightTrimPosition(@SqlType("varchar(x)") Slice slice, int[] codePointsToTrim)
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: @SqlType - ditto.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 6]
```diff
@@ -448,6 +449,88 @@ public static Slice trim(@SqlType("varchar(x)") Slice slice)
...
+
+    @Description("remove the longest string containing only given characters from the beginning and end of a string")
+    @ScalarFunction("trim")
+    @LiteralParameters("x")
+    @SqlType("varchar(x)")
+    public static Slice trim(@SqlType("varchar(x)") Slice slice, @SqlType(StandardTypes.VARCHAR) Slice charactersToTrim)
+    {
+        int[] codePointsToTrim = toCodePoints(charactersToTrim);
+        int start = leftTrimPosition(slice, codePointsToTrim);
+        if (start == slice.length()) {
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: Isn't this check redundant?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 7]
```diff
@@ -469,9 +552,9 @@ public static Slice upper(@SqlType("varchar(x)") Slice slice)
     private static Slice pad(Slice text, long targetLength, Slice padString, int paddingOffset)
     {
         checkCondition(
-            0 <= targetLength && targetLength <= Integer.MAX_VALUE,
```
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: Unrelated changes

**Comment by fiedukow** [comment]
LGTM % comments:
- SqlType annotation
- Second argument input type

Nothing crucial in other comments ;)

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java (hunk 6 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: it is optimization ;)

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java (hunk 1 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: I based that on PSQL version of `ltrim`. They have used `TEXT` so it like ours `varchar`. I think that `varchar` is more convenient for user, while `array(char(1))` is more semantically correct.  I vote to keep it as it is.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java (hunk 3 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: I knew you wouldn't like that ;)

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java (hunk 7 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: Extracted to separate commit.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java (hunk 4 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: It is not as simple as it sounds. That was my first idea. SliceUtf8 trim functions are optimized for whitespace characters. It is using static array of whitespace code points. So anyway I would need to create separate functions there (not much common code).

I could put these functions there, but I am not sure if I should. Anyway I can do that as separate pull requests as SliceUtf8 belongs to separate project.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java (hunk 3 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: I knew that it is a review-trap ;-)

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java (hunk 6 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: Well I guess so. But are you sure if this isn't slower in general case? :P
For me this optimization is to small to complicate the code, but it's your call ;)

**Comment by martint** [comment]
@haozhun, can you review?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 8]
```diff
@@ -448,6 +449,90 @@ public static Slice trim(@SqlType("varchar(x)") Slice slice)
...
+
+    private static int rightTrimPosition(Slice slice, int[] codePointsToTrim)
+    {
+        int end = slice.length();
+        while (end > 0) {
+            boolean found = false;
+            int codePoint = SliceUtf8.getCodePointBefore(slice, end);
+            for (int codePointToTrim : codePointsToTrim) {
+                if (codePoint == codePointToTrim) {
+                    end--;
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: This does not work with non latin1 characters.

**Comment by haozhun** [comment]
Our experience is that throwing on invalid UTF8 is a bad idea because invalid UTF8 entries usually exist. Therefore, we decided that for String functions in Presto, we use garbage-in-garbage-out policy. That is, we never throw for data that would normally come from a table.

In addition to this, we make judgemental calls on whether an argument of a function comes from table or query. For example, for `trim(slice, charactersToTrim)`, we would throw if `charactersToTrim` is invalid, but would do garbage-in-garbage-out when `slice` is invalid.

**Comment by haozhun** [comment]
Add unicode tests

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java (hunk 8 above)
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: Add tests that contain non-latin1 characters

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java (hunk 3 above)
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: I don't generally have a problem with labels in Java. But I think in this particular case, I think it's better to write it with a boolean flag. Having an unconditional `break` at the end of a loop, and then skipping over it with a labeled `continue` seems convoluted.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 9]
```diff
@@ -448,6 +449,90 @@ public static Slice trim(@SqlType("varchar(x)") Slice slice)
...
+
+    private static int leftTrimPosition(Slice slice, int[] codePointsToTrim)
+    {
+        int start = 0;
+        SLICE_LOOP:
+        while (start < slice.length()) {
+            int codePoint = getCodePointAt(slice, start);
+            for (int codePointToTrim : codePointsToTrim) {
+                if (codePoint == codePointToTrim) {
+                    start += lengthOfCodePoint(start);
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: wrong argument

**Comment by haozhun** [comment]
I have a few comments. One of them (the one about labeled continue) is under an old thread.

**Comment by kokosing** [comment]
@haozhun Thanks for comments, I have just applied them. Please take another look.

**Comment by haozhun** [comment]
Thank you for the update.

A few comments:
* We prefer not to have non-latin characters in source code. Please turn those into `\uxxxx`. IDE should be able to do it for you.
* The implementation looks good. But tests are lacking, especially for `rightTrimPosition`. That are many uncovered branches in this piece of logic-heavy code. I would either add tests that directly tests the function. Or add additional tests for `trim` to try exercising those. I think adding direct tests are probably better in this case.
* There is clearly an optimization opportunity here for preprocessing the `charactersToTrim`, which is possible with today's Presto. But that would make the code more complicated. (See how we did it for regex functions.) For now, I would just add a `TODO` to the code. We can delay doing that until we find it necessary. I may provide more update on this later.

Feel free to merge fixup commits into their original ones for this pull request.

Additionally,

* I'm discussing with @martint and @dain about moving some of your code to `SliceUtf8` and/or opening up some of the functions there to minimize code copying. For any code we decide to move, I can make the changes on the `slice` side for you.

**Comment by haozhun** [comment]
Please hold off addressing my comments. I'll add the necessary functions in airlift slice.

**Comment by haozhun** [comment]
@kokosing The necessary implementations are now available in `slice`. Can you please update this pull request to utilize the implementations provided there? If you can't find the methods you need, this commit (https://github.com/haozhun/presto/commit/215efcfaf5aead1b73b3edbcece7fbebe4d5440c) may help you.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java [hunk 10]
```diff
@@ -380,6 +380,102 @@ public void testTrim()
...
+
+        // non latin characters
+        assertFunction("RTRIM('źółć', 'ćł')", createVarcharType(4), "źó");
+
+        // invalid utf-8 characters
+        assertFunction("CAST(RTRIM(utf8(from_hex('81')), ' ') AS VARBINARY)", VARBINARY, varbinary(0x81));
+        assertFunction("CAST(RTRIM(CONCAT(utf8(from_hex('81')), ' '), ' ') AS VARBINARY)", VARBINARY, varbinary(0x81));
+        assertFunction("CAST(RTRIM(CONCAT(' ', utf8(from_hex('81'))), ' ') AS VARBINARY)", VARBINARY, varbinary(' ', 0x81));
+        assertFunction("CAST(RTRIM(CONCAT(' ', utf8(from_hex('81')), ' '), ' ') AS VARBINARY)", VARBINARY, varbinary(' ', 0x81));
+        assertInvalidFunction("RTRIM('źółć', utf8(from_hex('81')))", createVarcharType(4), "Invalid UTF-8 encoding");
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java]: When you're testing invalid second argument, it doesn't really matter what you provide in the first argument. I'll just use something simple (instead of non-ASCII characters)

**Comment by haozhun** [comment]
@kokosing Back to you for comments.

**Comment by kokosing** [comment]
Thanks. I will update the PR next week.

pt., 24.06.2016, 01:14 użytkownik Haozhun Jin <notifications@github.com>
napisał:

**Comment by kokosing** [comment]
@haozhun back to you ;) Comments were applied.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 11]
```diff
@@ -448,6 +452,60 @@ public static Slice trim(@SqlType("varchar(x)") Slice slice)
...
+        return codePoints;
+    }
+
+    private static int safeCountCodePoints(Slice slice)
+    {
+        int codePoints = 0;
+        for (int position = 0; position < slice.length(); ) {
+            int codePoint = tryGetCodePointAt(slice, position);
+            if (codePoint < 0) {
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Invalid UTF-8 encoding");
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: `"Invalid UTF-8 encoding in characters to trim: " + slice.toStringUtf8()`. Change the error messages in tests accordingly. `toStringUtf8()` turns invalid characters into replacement characters.

**Comment by haozhun** [comment]
Looks good otherwise. I'll merge after the minor comment is addressed.

**Comment by kokosing** [comment]
Rebased and comment applied.

**Comment by haozhun** [comment]
Merged #5327. Thank you!

I made two additional changes before I merge:
* I removed the reference to this pull request from the commit message
* I reordered the functions in the 2nd commit (when the functions were introduced) so that the 3rd commit does not need to reorder them.
